### PR TITLE
Rust: Type inference for pattern matching

### PR DESCRIPTION
**Pull request by hvitved**

### CODE REVIEW

File: rust/ql/lib/codeql/rust/internal/TypeInference.qll [hunk 1]
```diff
@@ -1178,6 +1198,120 @@ private Type inferIndexExprType(IndexExpr ie, TypePath path) {
...
+ */
+private module StructPatMatchingInput implements MatchingInputSig {
+  class DeclarationPosition = StructExprMatchingInput::DeclarationPosition;
+
+  class Declaration = StructExprMatchingInput::Declaration;
+
+  class AccessPosition = DeclarationPosition;
+
+  class Access extends StructPat {
+    Type getTypeArgument(TypeArgumentPosition apos, TypePath path) { none() }
```
Comment (aibaars) [inline rust/ql/lib/codeql/rust/internal/TypeInference.qll:1225]: Why is this `none()`?

File: rust/ql/test/library-tests/type-inference/main.rs [hunk 2]
```diff
@@ -2383,15 +2383,15 @@ pub mod pattern_matching {
...
         if let MyRecordStruct { value1, value2 } = my_record_struct {
-            let x = value1; // $ MISSING: type=x:i32
-            let y = value2; // $ MISSING: type=y:bool
+            let x = value1; // $ type=x:i32
+            let y = value2; // $ type=y:bool
             ();
         }
 
         let my_tuple_struct = MyTupleStruct(42, false);
         if let MyTupleStruct(value1, value2) = my_tuple_struct {
```
Comment (aibaars) [inline rust/ql/test/library-tests/type-inference/main.rs:2418]: There are a bunch of other pattern [types](https://github.com/rust-lang/rust-analyzer/blob/master/crates/syntax/rust.ungram#L688-L704), lets add some tests for those too. No need to fully implement them all, having some `MISSING` annotations is fine.

Let's also add some more tests with:
* some more deeply nested patterns
* use of `ref` keyword: `let Some(ref x) = ...`
* a `union` type

**Review by aibaars — COMMENTED** [review]
Very nice! Are you planning to add pattern matching on tuple types as well? Or is that best left as a follow-up?

**Comment by hvitved** [comment]
> Very nice! Are you planning to add pattern matching on tuple types as well? Or is that best left as a follow-up?

I was thinking we could do that once we even support tuple types.

File: rust/ql/test/library-tests/type-inference/main.rs (hunk 2 above)
Comment (hvitved) [reply rust/ql/test/library-tests/type-inference/main.rs:2418]: Good idea with nested patterns and `ref`; we don't handle union types yet, so I'll not add those here.

File: rust/ql/lib/codeql/rust/internal/TypeInference.qll (hunk 1 above)
Comment (hvitved) [reply rust/ql/lib/codeql/rust/internal/TypeInference.qll:1225]: Since https://github.com/github/codeql/pull/19847, `getTypeArgument` is only used for type arguments supplied to functions, not to their defining type, i.e. it is used for `foo::<i32>(...)` but not for `Bar::<i32>::foo(...)`.

Since explicit type arguments in pattern matching corresponds to the latter, it is `none()` here, and instead handled in `getInferredType` below. I'll add some tests for this.

Comment (hvitved) [reply rust/ql/lib/codeql/rust/internal/TypeInference.qll:1225]: Adding test, of course, revealed some corner cases that were not handled, I'm gonna do that on this PR as well.

File: rust/ql/test/library-tests/type-inference/main.rs (hunk 2 above)
Comment (aibaars) [reply rust/ql/test/library-tests/type-inference/main.rs:2418]: I just asked co-pilot to generate sample test code that covers all the cases for `Pat` from the ungrammar. This is the result https://github.com/hvitved/codeql/pull/8 . 

I had to drop the sample code about "const block patterns" because that feature has been removed from Rust.

**Review by aibaars — DISMISSED** [review]
LGTM

File: rust/ql/test/library-tests/type-inference/main.rs (hunk 2 above)
Comment (hvitved) [reply rust/ql/test/library-tests/type-inference/main.rs:2418]: Could we perhaps add those tests follow-up?

**Review by geoffw0 — DISMISSED** [review]
LGTM as well.

I also looked at the DCA run, it shows a pretty substantial improvement on missing call targets, and some new results for `rust/access-after-lifetime-ended` (probably as we're identifying more sinks correctly). :+1:

File: rust/ql/test/library-tests/type-inference/main.rs (hunk 2 above)
Comment (aibaars) [reply rust/ql/test/library-tests/type-inference/main.rs:2418]: Sure, just wanted to let you know, so you wouldn't write them manually.

Comment (hvitved) [reply rust/ql/test/library-tests/type-inference/main.rs:2418]: Thanks :-)
